File: modulo04-agentes-autonomos/agent-memory-java/src/main/java/com/iadeva/memory/agent/MemoryAwareAgent.java
package com.iadeva.memory.agent;

// Equivalente ao memory_aware_agent.py da aula 13 — agente que integra os 4 tipos de memória no ciclo

import com.iadeva.memory.engine.ReflectionEngine;
import com.iadeva.memory.memory.ContextualMemory;
import com.iadeva.memory.memory.EpisodicMemory;
import com.iadeva.memory.memory.LongTermMemory;
import com.iadeva.memory.memory.ShortTermMemory;
import com.iadeva.memory.model.AgentRunResult;
import com.iadeva.memory.model.Episode;
import com.iadeva.memory.model.Lesson;
import com.iadeva.memory.model.MemoryFact;
import org.springframework.ai.chat.client.ChatClient;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Agente com consciência de memória: integra os 4 tipos de memória no ciclo de execução.
 * Fluxo: contexto semântico → fatos → episódios → execução enriquecida → persistência → reflexão.
 */
@Component
public class MemoryAwareAgent {

    private final ApplicationContext applicationContext;
    private final LongTermMemory longTermMemory;
    private final EpisodicMemory episodicMemory;
    private final ContextualMemory contextualMemory;
    private final ReflectionEngine reflectionEngine;
    private final ChatClient chatClient;

    public MemoryAwareAgent(
            ApplicationContext applicationContext,
            LongTermMemory longTermMemory,
            EpisodicMemory episodicMemory,
            ContextualMemory contextualMemory,
            ReflectionEngine reflectionEngine,
            ChatClient.Builder chatClientBuilder) {
        this.applicationContext = applicationContext;
        this.longTermMemory = longTermMemory;
        this.episodicMemory = episodicMemory;
        this.contextualMemory = contextualMemory;
        this.reflectionEngine = reflectionEngine;
        this.chatClient = chatClientBuilder.build();
    }

    /**
     * Executa o agente com contexto enriquecido por todos os 4 tipos de memória.
     *
     * @param input pergunta ou instrução do usuário
     * @return resultado com output e métricas de uso de memória
     */
    public AgentRunResult run(String input) {
        // Obtém instância prototype isolada para esta execução
        ShortTermMemory shortTerm = applicationContext.getBean(ShortTermMemory.class);
        List<String> steps = new ArrayList<>();

        try {
            // PASSO 1: Busca contexto semântico via ContextualMemory
            steps.add("Buscando contexto semântico via embeddings");
            List<ContextualMemory.EmbeddingEntry> contextoSemantico = buscarContexto(input);
            shortTerm.put("contexto_semantico", contextoSemantico);

            // PASSO 2: Recupera fatos relevantes de LongTermMemory
            steps.add("Recuperando fatos da memória de longo prazo");
            List<MemoryFact> fatos = longTermMemory.getFacts();
            shortTerm.put("fatos", fatos);

            // PASSO 3: Consulta episódios recentes de EpisodicMemory
            steps.add("Consultando episódios recentes");
            List<Episode> episodiosRecentes = episodicMemory.getRecentEpisodes(3);
            shortTerm.put("episodios_recentes", episodiosRecentes);

            // PASSO 4: Executa com contexto enriquecido via ChatClient
            steps.add("Executando com contexto enriquecido");
            String promptEnriquecido = construirPromptEnriquecido(input, fatos, episodiosRecentes, contextoSemantico);
            String output = chatClient.prompt()
                    .user(promptEnriquecido)
                    .call()
                    .content();
            shortTerm.put("output", output);

            // PASSO 5: Persiste episódio + extrai lições via ReflectionEngine
            steps.add("Persistindo episódio e extraindo lições");
            Episode episodio = new Episode(
                    UUID.randomUUID().toString(),
                    input,
                    List.copyOf(steps),
                    output,
                    List.of(),
                    Instant.now()
            );
            episodicMemory.addEpisode(episodio);

            // Armazena o input no índice semântico para futuras execuções
            contextualMemory.store(input, Map.of("tipo", "input", "timestamp", Instant.now().toString()));

            // Reflexão: extrai lições se houver aprendizado relevante
            List<Lesson> licoes = reflectionEngine.reflect(episodio);
            if (!licoes.isEmpty()) {
                steps.add("Lições extraídas: " + licoes.size());
            }

            return new AgentRunResult(
                    output,
                    true,
                    fatos.size(),
                    episodiosRecentes.size()
            );

        } finally {
            // Garante limpeza da memória de curto prazo ao fim da execução
            shortTerm.clear();
        }
    }

    // Busca contexto semântico — retorna lista vazia se memória contextual ainda não tem entradas
    private List<ContextualMemory.EmbeddingEntry> buscarContexto(String input) {
        try {
            return contextualMemory.search(input, 3);
        } catch (Exception e) {
            // Memória contextual pode estar vazia na primeira execução
            return List.of();
        }
    }

    // Monta prompt enriquecido com todos os contextos disponíveis
    private String construirPromptEnriquecido(
            String input,
            List<MemoryFact> fatos,
            List<Episode> episodios,
            List<ContextualMemory.EmbeddingEntry> contexto) {

        StringBuilder sb = new StringBuilder();

        // Contexto semântico relevante
        if (!contexto.isEmpty()) {
            sb.append("=== CONTEXTO SEMÂNTICO RELEVANTE ===\n");
            contexto.forEach(e -> sb.append("- ").append(e.content()).append("\n"));
            sb.append("\n");
        }

        // Fatos da memória de longo prazo
        if (!fatos.isEmpty()) {
            sb.append("=== FATOS CONHECIDOS ===\n");
            fatos.forEach(f -> sb.append("- ").append(f.content())
                    .append(" [fonte: ").append(f.source()).append("]\n"));
            sb.append("\n");
        }

        // Histórico de episódios recentes
        if (!episodios.isEmpty()) {
            sb.append("=== HISTÓRICO RECENTE ===\n");
            episodios.forEach(e -> sb.append("Input anterior: ").append(e.input())
                    .append(" → Resultado: ").append(resumir(e.outcome())).append("\n"));
            sb.append("\n");
        }

        sb.append("=== TAREFA ATUAL ===\n").append(input);
        return sb.toString();
    }

    private String resumir(String texto) {
        if (texto == null) return "sem resultado";
        return texto.length() > 100 ? texto.substring(0, 100) + "..." : texto;
    }
}

File: modulo04-agentes-autonomos/agent-memory-go/agent/memory_aware_agent.go
// Package agent implementa o agente com consciência de memória — equivalente
// a MemoryAwareAgent.java (aula 13 do curso).
package agent

import (
	"context"
	"fmt"
	"strings"
	"time"

	"agent-memory/engine"
	"agent-memory/memory"
	"agent-memory/model"

	"github.com/google/uuid"
)

// ChatClient é a única operação de que MemoryAwareAgent depende para gerar
// a resposta enriquecida.
type ChatClient interface {
	Chat(ctx context.Context, userPrompt string) (string, error)
}

// MemoryAwareAgent integra os 4 tipos de memória no ciclo de execução:
// contexto semântico → fatos → episódios → execução enriquecida →
// persistência → reflexão.
type MemoryAwareAgent struct {
	LongTerm   *memory.LongTermMemory
	Episodic   *memory.EpisodicMemory
	Contextual *memory.ContextualMemory
	Reflection *engine.ReflectionEngine
	Client     ChatClient
}

// Run executa o agente com contexto enriquecido por todos os 4 tipos de
// memória — equivalente a MemoryAwareAgent.run.
func (a *MemoryAwareAgent) Run(ctx context.Context, input string) (model.AgentRunResult, error) {
	shortTerm := memory.NewShortTermMemory()
	var steps []string
	defer shortTerm.Clear()

	// PASSO 1: contexto semântico via ContextualMemory
	steps = append(steps, "Buscando contexto semântico via embeddings")
	semanticContext := a.searchContext(ctx, input)
	shortTerm.Put("contexto_semantico", semanticContext)

	// PASSO 2: fatos relevantes de LongTermMemory
	steps = append(steps, "Recuperando fatos da memória de longo prazo")
	facts := a.LongTerm.GetFacts()
	shortTerm.Put("fatos", facts)

	// PASSO 3: episódios recentes de EpisodicMemory
	steps = append(steps, "Consultando episódios recentes")
	recentEpisodes := a.Episodic.GetRecentEpisodes(3)
	shortTerm.Put("episodios_recentes", recentEpisodes)

	// PASSO 4: execução com contexto enriquecido
	steps = append(steps, "Executando com contexto enriquecido")
	enrichedPrompt := buildEnrichedPrompt(input, facts, recentEpisodes, semanticContext)
	output, err := a.Client.Chat(ctx, enrichedPrompt)
	if err != nil {
		return model.AgentRunResult{}, err
	}
	shortTerm.Put("output", output)

	// PASSO 5: persiste episódio + extrai lições via ReflectionEngine
	steps = append(steps, "Persistindo episódio e extraindo lições")
	episode := model.Episode{
		ID:               uuid.NewString(),
		Input:            input,
		Steps:            append([]string{}, steps...),
		Outcome:          output,
		LessonsExtracted: nil,
		Timestamp:        time.Now(),
	}
	a.Episodic.AddEpisode(episode)

	// Armazena o input no índice semântico para futuras execuções.
	a.Contextual.Store(ctx, input, map[string]any{"tipo": "input", "timestamp": time.Now().Format(time.RFC3339)})

	lessons := a.Reflection.Reflect(ctx, episode)
	if len(lessons) > 0 {
		steps = append(steps, fmt.Sprintf("Lições extraídas: %d", len(lessons)))
	}

	return model.AgentRunResult{
		Output:             output,
		MemoryUsed:         true,
		FactsRetrieved:     len(facts),
		EpisodesConsidered: len(recentEpisodes),
	}, nil
}

// searchContext busca contexto semântico — retorna vazio se a memória
// contextual ainda não tem entradas ou a busca falhar (primeira execução).
func (a *MemoryAwareAgent) searchContext(ctx context.Context, input string) []memory.EmbeddingEntry {
	results, err := a.Contextual.Search(ctx, input, 3)
	if err != nil {
		return nil
	}
	return results
}

func buildEnrichedPrompt(input string, facts []model.MemoryFact, episodes []model.Episode, context []memory.EmbeddingEntry) string {
	var sb strings.Builder

	if len(context) > 0 {
		sb.WriteString("=== CONTEXTO SEMÂNTICO RELEVANTE ===\n")
		for _, e := range context {
			sb.WriteString("- " + e.Content + "\n")
		}
		sb.WriteString("\n")
	}

	if len(facts) > 0 {
		sb.WriteString("=== FATOS CONHECIDOS ===\n")
		for _, f := range facts {
			sb.WriteString(fmt.Sprintf("- %s [fonte: %s]\n", f.Content, f.Source))
		}
		sb.WriteString("\n")
	}

	if len(episodes) > 0 {
		sb.WriteString("=== HISTÓRICO RECENTE ===\n")
		for _, e := range episodes {
			sb.WriteString(fmt.Sprintf("Input anterior: %s → Resultado: %s\n", e.Input, summarize(e.Outcome)))
		}
		sb.WriteString("\n")
	}

	sb.WriteString("=== TAREFA ATUAL ===\n")
	sb.WriteString(input)
	return sb.String()
}

func summarize(text string) string {
	if text == "" {
		return "sem resultado"
	}
	if len(text) > 100 {
		return text[:100] + "..."
	}
	return text
}
